row.py

```diff
@@ -558,3 +558,83 @@ def summarize_run(folder, run_name):
             "unrecognized characters": len(frame.loc[(frame["Empty Text"] > 0), :].index),
         },
     )
+
+
+def build_mosaic_image(images, object_name):
+    """build a mosaic image from a list of cv2 images
+
+    Args:
+        images (list): list of cv2 images to mosaic together
+        object_name (str): the name of the image object (original filename)
+
+    Returns:
+        mosaic_image (np.ndarray): composite mosaic of smaller images
+    """
+    max_width = 0
+    buffer = 5
+
+    #: Loop through all images to get sizes, save largest
+    for img in images:
+        if img.shape[1] > max_width:
+            max_width = img.shape[1]
+
+    #: Set up parameters for mosaic, calculate number of cols/rows
+    number_images = len(images)
+    number_columns = math.floor(math.sqrt(number_images))
+    number_rows = math.ceil(number_images / number_columns)
+
+    logging.info("mosaicking %i images into %i column by %i row grid", number_images, number_columns, number_rows)
+
+    #: Build mosaic image with white background
+    tile_width = max_width + 2 * buffer
+    total_height = tile_width * number_rows
+    total_width = tile_width * number_columns
+    mosaic_image = np.zeros((total_height, total_width, 3), dtype=np.uint8)
+    mosaic_image[:, :] = (255, 255, 255)
+
+    if total_height * total_width > 40_000_000:
+        logging.info('mosaic image size is too large: "%s"', object_name)
+
+        return None
+
+    i = 0
+    for img in images:
+        #: Resize all images by inserting them into the same template tile size
+        [img_height, img_width, _] = img.shape
+        buffered_image = np.zeros((tile_width, tile_width, 3), np.uint8)
+        buffered_image[:, :] = (255, 255, 255)
+        buffered_image[buffer : buffer + img_height, buffer : buffer + img_width] = img.copy()
```

Comment:
**thought**: i don't think this is necessary. I would assume the OCR will have the same results regardless of size no? 

Is this mainly to avoid tiling collisions? I worry a little that resizing will affect the quality which could affect the OCR. A lot of the images are a lossy format.

Reply:
This is creating a blank canvas (white pixels) and then inserting the existing image into it.  It allows the tiling to be even and cleanly stitched together.  It's not actually changing resolution or anything, just inserting the smaller cropped images into an array that matches the largest cropped images size. Above this on [line 574](https://github.com/agrc/udot-parcel-ml/blob/refactor/mosaic-images/row.py#L574)  I do create a 5 pixel border around the largest image, just to make sure the tiling is a little cleaner and separated.

Reply:
With equal sizing for each "tile" it's much easier to calculate where the tile should be placed in the large mosaic.

Reply:
So the circles themselves remain the same dimensions as the original? 

It's not that difficult to get the max height for the circles and use that as the row height if the circles have different dimensions.

Reply:
Yes, the circles are the same dimensions.  I'm going for a nice checkerboard mosaic where all columns/rows are equal width/height.

Reply:
So then we are resizing the circles

Reply:
ok i get it.